File: apps/server/src/storage/paths.ts
// Where perch keeps everything on disk. One directory holds all of it (default ~/.perch, override
// with PERCH_HOME): connections.json (0600), settings.json, history.jsonl, server.json —
// plus queries/, the workspace folder created on first run. Nothing perch writes lives anywhere
// else.

import { promises as fs } from "node:fs";
import os from "node:os";
import path from "node:path";

export function configDir(): string {
  return process.env.PERCH_HOME ?? path.join(os.homedir(), ".perch");
}

/** The workspace root created on first run: ~/.perch/queries, where .sql files live by default. */
export function defaultQueriesDir(): string {
  return path.join(configDir(), "queries");
}

/** Creates perch home if it is missing and returns its path. */
export async function ensureDir(): Promise<string> {
  const dir = configDir();
  await fs.mkdir(dir, { recursive: true, mode: 0o700 });
  return dir;
}

/** Absolute path of one file inside perch home. */
export function configFile(name: string): string {
  return path.join(configDir(), name);
}

export const CONNECTIONS_FILE = "connections.json";
export const SETTINGS_FILE = "settings.json";
export const HISTORY_FILE = "history.jsonl";
export const SERVER_INFO_FILE = "server.json";

File: apps/server/storage/paths.go
// Package storage owns everything perch keeps on disk. One directory holds all of it (default
// ~/.perch, override with PERCH_HOME): connections.json (0600), settings.json, history.jsonl,
// server.json, update.json, plus queries/, the workspace created on first run.
package storage

import (
	"os"
	"path/filepath"
)

const (
	ConnectionsFile = "connections.json"
	SettingsFile    = "settings.json"
	HistoryFile     = "history.jsonl"
	ServerInfoFile  = "server.json"
	UpdateCheckFile = "update.json"
	QueriesDirName  = "queries"
)

func ConfigDir() string {
	if dir := os.Getenv("PERCH_HOME"); dir != "" {
		return dir
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ".perch"
	}
	return filepath.Join(home, ".perch")
}

// DefaultQueriesDir is the workspace root created on first run, where .sql files live by default.
func DefaultQueriesDir() string {
	return filepath.Join(ConfigDir(), QueriesDirName)
}

func EnsureDir() (string, error) {
	dir := ConfigDir()
	if err := os.MkdirAll(dir, 0o700); err != nil {
		return "", err
	}
	return dir, nil
}

func ConfigFile(name string) string {
	return filepath.Join(ConfigDir(), name)
}
